src/System.Text.Primitives/System/Text/Encoders/Utf8Util.Helpers.cs

```diff
@@ -0,0 +1,441 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+//
+// This file contains utility methods shared by the UTF-8 workhorse methods.
+//
+
+namespace System.Buffers.Text
+{
+    internal static partial class Utf8Util
```

Comment:
All the APIs seem to related to UTF8 validation. Should this type be called Utf8ValidationHelpers?

Reply:
There will also be UTF8-to-UTF16 code in this same static class at some point. But you’re right, the naming is horrible. Perhaps simply Utf8Helpers (drop the “validation”) would work?